[Dynamo] support str.format

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99207

Allow `str.format` on constants, tensors, lists without introducing a graph break in Dynamo.

The intended use case is printing. Note that we do not emit code here, but instead burn in the result of `str.format`. Discussed offline with @yanboliang.

Differential Revision: [D45016752](https://our.internmc.facebook.com/intern/diff/D45016752/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov @soumith @desertfire

avikchaudhuri committed:
[Dynamo] support str.format

Allow `str.format` on constants, tensors, lists without introducing a graph break in Dynamo.

The intended use case is printing. Note that we do not emit code here, but instead burn in the result of `str.format`. Discussed offline with @yanboliang.

Differential Revision: [D45016752](https://our.internmc.facebook.com/intern/diff/D45016752/)

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1172,6 +1172,25 @@ def call_id(self, tx, *args):
+            return value.as_python_constant()
+        elif isinstance(value, variables.TensorVariable):
+            return value.get_real_value()
+        elif isinstance(value, variables.BaseListVariable):
```
I think this has been covered by ```value.is_python_constant()```.

avikchaudhuri replied:
I don't think so, `is_python_constant` will fail for a list of tensors, I think.

yanboliang replied:
```is_python_constant``` won't fail, it returns ```False```.
https://github.com/pytorch/pytorch/blob/efc3887ea508b3cfd94603fd8afe4e8cf6dce7b7/torch/_dynamo/variables/base.py#L152-L157

yanboliang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1172,6 +1172,25 @@ def call_id(self, tx, *args):
+    def _raw_value_to_format(self, value):
+        if value.is_python_constant():
+            return value.as_python_constant()
+        elif isinstance(value, variables.TensorVariable):
```
It prints FakeTensor stuff, I don't know how much value it is for end users?

avikchaudhuri replied:
Not sure about how much value it is for end users, but what do you propose to do here instead? Something like <tensor> without details? 

Note that the motivating use case does have a tensor in the format output, so we need to support it at least to the extent that it doesn't crash.

jansel replied:
If the goal is to turn printing a Tensor into a no-op, then `<tensor>` might be better -- though since that is changing the semantics of the original program it should be behind a flag like `ignore_tensor_prints`.

yanboliang commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1172,6 +1172,25 @@ def call_id(self, tx, *args):
         else:
             unimplemented(f"call_id with args {args}")
 
+    def _raw_value_to_format(self, value):
```
I think another way for this is just implement ```__str__``` for every variable, most of the case you can just call ```.as_python_constant()```

avikchaudhuri replied:
Well yeah, except when they are containers and tensors. I'd expect that we'll also support dicts, e.g., and that's easy to add. There seem to be many other variable kinds, though. Do you suggest I support all of them in one go here?

yanboliang replied:
I think you can add ```__str__``` to base class ```VariableTracer```, and return ```as_python_constant``` if it is python constant. It can cover most of the cases, for the other variables, just add the separately.

avikchaudhuri committed:
Update on "[Dynamo] support str.format"


Allow `str.format` on constants, tensors, lists without introducing a graph break in Dynamo.

The intended use case is printing. Note that we do not emit code here, but instead burn in the result of `str.format`. Discussed offline with yanboliang.

Differential Revision: [D45016752](https://our.internmc.facebook.com/intern/diff/D45016752/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

jansel commented:
The ConstantVariable part of this will fix an issue with einops_ext, so we should land that part without a flag.  Only the changing semantics of tensor printing needs a flag.
